src/markov/sean.clj

```diff
@@ -0,0 +1,98 @@
+(ns markov.sean
+  (:gen-class))
+
+;; Generating a markov dictionary from a string of text
+;; e.g. { :first-words ["The" "An"] :last-words ["dog"] :dict { "The" ["quick" "large"] "fox" ["jumped"] ... } }
+
+(defn zip-pairs
+  "From a list, generates a list of sequential pairs ([1 2 3] becomes [[1 2] [2 3]])"
+  [list]
+  (map vector list (next list)))
+
+(defn build-dict
+  "Builds the graph of words in a markov dictionary, given an ordered list of words"
+  [words]
+  (reduce (fn [dict, current-pair]
+            (let [current-word (current-pair 0)
+                  next-word (current-pair 1)
+                  words-after-current (get dict current-word)]
+              (if words-after-current
+                (update dict current-word conj next-word) 
+                (merge dict {current-word [next-word]}))))
+          {} (zip-pairs words)))
+
+(defn split-into-words
+  [text]
+  (clojure.string/split text #" "))
+
+(defn get-first-words
+  "From a list of sentences, returns a list of the first words"
+  [sentences]
+  (map (fn [sentence]
+         (first (split-into-words sentence))) sentences))
+
+(defn get-last-words
+  "From a list of sentences, returns a list of the last words"
+  [sentences]
+  (map (fn [sentence]
+         (last (split-into-words sentence))) sentences))
+
+(defn markov-generate-dict
+  "Builds a markov dict for some text, including a list of first and last words."
+  [text]
+  (let [sentences (clojure.string/split text #"\. ") ; TODO: handle '?'
+        words (mapcat split-into-words sentences)
+        first-words (get-first-words sentences)
+        last-words (get-last-words sentences)]
+    (hash-map :first-words first-words :last-words last-words :dict (build-dict words))))
+
+;; Generating text from a markov dictionary
+
+(defn next-word
+  "Gets the next word, given a word and a markov dict."
+  [word dict]
+  (rand-nth (get (:dict dict) word)))
+
+(defn recursively-construct-sentence
+  "Adds a word to an array of words from the markov dict, if possible."
+  [sentence-array dict]
+  (let [current-word (last sentence-array)]
+    (if (.contains (:last-words dict) current-word) ; `contains?` is a well-named method and clojure is a good language
```

Comment:
`contains?` sucks. The argument has been made that it should be changed. And counter argued that Clojure should never break people's code. Either way it's a shame, and I think they got it wrong.

Reply:
If you had `:last-words` as a set (like `#{"end" "fin"}` then you could check that `current-word` is in that set by using it as a function:

``` clojure
(def last-words #{"end" "fin"})

(last-words "end") # => "end"
(last-words "potato") # => nil
```

Reply:
Also if you like you can destructure maps (and other associative things like records) using `:keys`. So in your argument list you could have `[sentence-array {:keys [last-words] :as dict}]` and then in the body of your function you could reference `last-words` and `dict`.

That in combination with the set thing would make this line:

```clojure
(if (last-words current-word)
```

Reading it now I'm not sure that with those choices of variable names it's immediately obvious what's going on, but it is shorter.